src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -2,60 +2,122 @@
 
-        # Equally corrupt head and tail
-        split_idx = num_negs // 2
+        # Equally corrupt all sides
+        split_idx = num_negs // len(self._corruption_indices)
 
         # Copy positive batch for corruption.
         # Do not detach, as no gradients should flow into the indices.
         negative_batch = positive_batch.clone()
 
         # Sample random entities as replacement
-        negative_entities = torch.randint(high=self.num_entities - 1, size=(num_negs,), device=positive_batch.device)
-
-        # Replace heads – To make sure we don't replace the head by the original value
-        # we shift all values greater or equal than the original value by one up
-        # for that reason we choose the random value from [0, num_entities -1]
-        filter_same_head = (negative_entities[:split_idx] >= positive_batch[:split_idx, 0])
-        negative_batch[:split_idx, 0] = negative_entities[:split_idx] + filter_same_head.long()
-        # Corrupt tails
-        filter_same_tail = (negative_entities[split_idx:] >= positive_batch[split_idx:, 2])
-        negative_batch[split_idx:, 2] = negative_entities[split_idx:] + filter_same_tail.long()
-
-        return negative_batch
+        if 0 in self._corruption_indices or 2 in self._corruption_indices:
+            negative_entities = torch.randint(
+                high=self.num_entities - 1,
+                size=(num_negs,),
+                device=positive_batch.device,
+            )
+
+        # Sample random relations as replacement, if requested
+        if 1 in self._corruption_indices:
+            negative_relations = torch.randint(
+                high=self.num_relations - 1,
+                size=(num_negs,),
+                device=positive_batch.device,
+            )
+
+        for index, start in zip(self._corruption_indices, range(0, num_negs, split_idx)):
+            stop = min(start + split_idx, num_negs)
+            if index == 1:
+                # Corrupt relations
+                negative_batch[start:stop, index] = negative_relations[start:stop]
+            else:
+                # Corrupt heads or tails
+                negative_batch[start:stop, index] = negative_entities[start:stop]
+
+            # To make sure we don't replace the {head, relation, tail} by the
+            # original value we shift all values greater or equal than the original value by one up
+            # for that reason we choose the random value from [0, num_{heads, relations, tails} -1]
+            if not self.filtered:
+                negative_batch[start:stop, index] += (
+                    negative_batch[start:stop, index] >= positive_batch[start:stop, index]
+                ).long()
+
+        # If filtering is activated, all negative triples that are positive in the training dataset will be removed
+        if self.filtered:
+            batch_filter = self.filter_negative_triples(negative_batch=negative_batch)
+            negative_batch = negative_batch[batch_filter]
```

Comment:
Since it is the same code for all samplers (if I see it correctly), consider moving this selection into the `filter_negative_triples` methods as well.